### PR TITLE
FAD-6306: Centralize most errors (4xx, etc)

### PR DESCRIPTION
**Pull request by bkemper**
_Refer to [FAD-6306](https://jira.int.messagesystems.com/browse/FAD-6306) for AC._

**Affected**

- Api Keys - creating, deleting, and updating
- Billing - updating contact info, updating plan, updating payment info
- Dashboard - verify sending domain
- Message events report including filtering
- In-App Support - creating
- Ip Pools - creating, deleting, updating
- Onboarding - create sending domain, create api key
- Profile (User) - updating info and password
- Recipient List - creating, deleting, searching, updating
- Sending Domains - creating, deleting, toggle default, updating, verifying, verifying with CName (the catch was removed, but a special case if `results.dns.cname_error` exists was left), verifying with DKIM, verifying with email
- SMTP - toggle engagement tracking
- Subaccounts - creating, updating
- Suppressions - deleting, get a list with filters, creating, uploading
- Templates - creating, deleting, previewing, publishing, updating
- Tracking domain - creating, deleting, updating, and verifying
- Users - deleting, inviting, updating
- Webhooks - creating, get refresh batches, deleting, showing (still redirects, but may remove), testing, updating

**Testing**

1. [Download Charles](https://www.charlesproxy.com/download/)
1. In "Proxy Settings", click "Use a dynamic port" in the "HTTP Proxy" panel.
1. In "Help" > "SSL Proxy", click "Install Charles Root Certificate".  [should open Keychain]
1. In Keychain, find the new certificate and double click it to open the info window
1. Expand "Trust" and change "When using this certificate" to "Always Trust"
1. Close the window to save the change
1. Start recording (the sequence tab is a better log than structure, especially with a filter)
1. Open production
1. Find a "CONNECT" API request (api.sparkpost.com), right click and click "SSL Proxying Enabled" from the menu
1. Reload the page [should see 200 GET requests with a path]
1. Open "Tools" > "Rewrite"
1. Import list of rewrites, [fad6306.xml.zip](https://github.com/SparkPost/2web2ui/files/1804295/fad6306.xml.zip)

**Todo**
- [x] Auto alert for SparkPost API errors
- [x] Auto alert for Zuora API errors
- [x] Remove manual alerting

### CODE REVIEW

File: src/components/reduxFormWrappers/FileFieldWrapper.js [hunk 1]
```diff
@@ -33,7 +33,7 @@ export default class FileFieldWrapper extends Component {
 
     return (
       <fieldset className={styles.Field}>
-        <Label id={input.id}>{label}{required && '*'}</Label>
+        <Label id={input.id}>{label}{required && ' *'}</Label>
```
Comment (bkemper) [inline src/components/reduxFormWrappers/FileFieldWrapper.js]: Quick win.

File: src/pages/reports/components/ReportOptions.js [hunk 2]
```diff
@@ -9,7 +9,6 @@ import { Grid, Button, Panel, Tag } from '@sparkpost/matchbox';
 import Typeahead from './Typeahead';
 import DateFilter from './DateFilter';
 import typeaheadCacheSelector from 'src/selectors/reportFilterTypeaheadCache';
-import { showAlert } from 'src/actions/globalAlert';
```
Comment (bkemper) [inline src/pages/reports/components/ReportOptions.js]: Clean-up.

File: src/pages/recipientLists/tests/CreatePage.test.js [hunk 3]
```diff
@@ -10,7 +10,7 @@ describe('CreatePage', () => {
     props = {
       createRecipientList: jest.fn(() => Promise.resolve()),
       showAlert: jest.fn(),
-      histort: { push: jest.fn() }
+      history: { push: jest.fn() }
```
Comment (bkemper) [inline src/pages/recipientLists/tests/CreatePage.test.js]: I have no idea how these tests passed.

File: src/pages/sendingDomains/components/EditBounce.js [hunk 4]
```diff
@@ -17,36 +17,26 @@ export class EditBounce extends Component {
...
     return update({
       id,
       subaccount: domain.subaccount_id,
       is_default_bounce_domain: !domain.is_default_bounce_domain
-    })
-      .catch((err) => {
-        showAlert({ type: 'error', message: `Failed to update default bounce option. ${err.message}` });
-        reset();
-      });
+    }).catch((err) => { reset(); });
```
Comment (bkemper) [inline src/pages/sendingDomains/components/EditBounce.js]: Need to refactor to avoid missing errors.

File: src/pages/users/CreatePage.js [hunk 5]
```diff
@@ -33,23 +33,6 @@ export class CreatePage extends Component {
           message: `Invitation sent to ${email}`
         });
         history.push('/account/users');
-      })
-      .catch((err) => {
-        const { response: { status, data }} = err;
-        let message = 'Unable to invite user.';
-
-        if (status === 400) {
```
Comment (bkemper) [inline src/pages/users/CreatePage.js]: Needs to be included when improving the error messages.

File: src/pages/webhooks/DetailsPage.js [hunk 6]
```diff
@@ -21,12 +21,9 @@ export class WebhooksDetails extends Component {
...
 
   componentDidMount() {
-    const { showAlert, getWebhook, match, history, subaccountId: subaccount } = this.props;
+    const { getWebhook, match, history, subaccountId: subaccount } = this.props;
 
-    getWebhook({ id: match.params.id, subaccount }).catch((err) => {
-      history.push('/webhooks/');
-      showAlert({ type: 'error', message: 'Unable to find webhook', details: err.message });
-    });
+    getWebhook({ id: match.params.id, subaccount }).catch((err) => { history.push('/webhooks/'); });
```
Comment (bkemper) [inline src/pages/webhooks/DetailsPage.js]: Refactor out this catch.

File: src/components/support/Support.js [hunk 7]
```diff
@@ -14,16 +13,11 @@ export class Support extends Component {
...
     const ticket = { subject, message };
-    return createTicket(ticket).catch((err) => {
-      showAlert({
-        type: 'error',
-        message: 'We were unable to submit your ticket. Please try again.'
-      });
-      throw err;
-    });
+
+    return createTicket(ticket).catch((err) => { throw err; });
```
Comment (bkemper) [inline src/components/support/Support.js]: Need to remove.

**Comment by jonambas** [comment]
Api keys looks good.
- [x] Api Keys - creating, deleting, and updating

Unrelated: looks like there's a small error in the empty state test - api error banner never shows
```js
// should check !error
show: count === 0 && !error,
```

File: src/actions/helpers/sparkpostApiRequest.js [hunk 8]
```diff
@@ -91,9 +91,8 @@ const sparkpostRequest = requestHelperFactory({
       meta
     });
 
-    if (response.status >= 500) {
-      dispatch(showAlert({ type: 'error', message: 'Something went wrong.', details: message }));
-    }
+    // auto alert all failures
+    dispatch(showAlert({ type: 'error', message: 'Something went wrong.', details: message }));
```
Comment (jonambas) [inline src/actions/helpers/sparkpostApiRequest.js]: Are we not implementing a silent flag?

File: src/pages/ipPools/EditPage.js [hunk 9]
```diff
@@ -57,11 +57,7 @@ export class EditPage extends Component {
...
         history.push('/account/ip-pools');
-      })
-      .catch(() => showAlert({
-        type: 'error',
-        message: `Unable to update IP pool ${id}.`
-      }));
+      });
   };
 
   onDeletePool = () => {
```
Comment (jonambas) [inline src/pages/ipPools/EditPage.js]: Need to remove showAlert from delete

File: src/pages/profile/ProfilePage.js [hunk 10]
```diff
@@ -25,25 +24,16 @@ export class ProfilePage extends Component {
...
       );
-
   }
 
   updatePassword = (values) => {
     const { username } = this.props.currentUser;
-    const { showAlert } = this.props;
     const { currentPassword, newPassword } = values;
 
     return this.props.confirmPassword(username, currentPassword)
```
Comment (jonambas) [inline src/pages/profile/ProfilePage.js]: No error shown if this first action fails

File: src/pages/profile/ProfilePage.js [hunk 11]
```diff
@@ -25,25 +24,16 @@ export class ProfilePage extends Component {
...
-      .then(() => this.props.updateUser(username, { password: newPassword }))
-      .catch((err) => showAlert({ type: 'error', message: 'Unable to update password' }));
+      .then(() => this.props.updateUser(username, { password: newPassword }));
   }
 
   render() {
```
Comment (jonambas) [inline src/pages/profile/ProfilePage.js]: TfaManager 
- if you hit enable and getSecret fails, alert is shown but UI shows loading forever (should maybe close the modal)

File: src/pages/ipPools/EditPage.js (hunk 9 above)
Comment (bkemper) [reply src/pages/ipPools/EditPage.js]: Addressed... 3005ee8f

File: src/pages/profile/ProfilePage.js (hunk 10 above)
Comment (bkemper) [reply src/pages/profile/ProfilePage.js]: Addressed... fcd4f62b

Comment (bkemper) [reply src/pages/profile/ProfilePage.js]: Not my favorite solution, but `confirmPassword` action creator isn't being used anywhere else.

File: src/pages/sendingDomains/components/EditBounce.js (hunk 4 above)
Comment (bkemper) [reply src/pages/sendingDomains/components/EditBounce.js]: Addressed... 4d6d6735

I slapped a TODO tag on it with a recommendation for how to remove it, but I'm not going to refactor it at this time.

File: src/pages/webhooks/DetailsPage.js (hunk 6 above)
Comment (bkemper) [reply src/pages/webhooks/DetailsPage.js]: I'm going to hold off on this as well.  I think this redirect logic should be abstracted into our routing layer.

File: src/components/support/Support.js (hunk 7 above)
Comment (bkemper) [reply src/components/support/Support.js]: Addressed... 0e78ad6c

File: src/actions/helpers/sparkpostApiRequest.js (hunk 8 above)
Comment (bkemper) [reply src/actions/helpers/sparkpostApiRequest.js]: Addressed... 887ed7bd

File: src/pages/profile/ProfilePage.js (hunk 10 above)
Comment (jonambas) [reply src/pages/profile/ProfilePage.js]: Hmm seems ok

File: src/actions/auth.js [hunk 12]
```diff
@@ -101,12 +102,17 @@ export function confirmPassword(username, password) {
...
+
         dispatch({
           type: 'CONFIRM_PASSWORD_FAIL',
           payload: {
             errorDescription
           }
         });
+
+        // To match sparkpostApiRequest behavior
+        dispatch(showAlert({ type: 'error', message: errorDescription }));
```
Comment (jonambas) [inline src/actions/auth.js]: May want to use `errorDescription` in the details key, and use a friendly message under `message`

dev tools request blocking produces a blank alert for me

File: src/actions/messageEvents.js [hunk 13]
```diff
@@ -2,7 +2,6 @@ import moment from 'moment';
 import config from 'src/config';
 import _ from 'lodash';
 import sparkpostApiRequest from 'src/actions/helpers/sparkpostApiRequest';
```
Comment (jonambas) [inline src/actions/messageEvents.js]: `getDocumentation` in this file can be silent - used only for tooltips

**Comment by jonambas** [comment]
sendingDomains `list` actions can be silent (i think)

File: src/actions/messageEvents.js (hunk 13 above)
Comment (bkemper) [reply src/actions/messageEvents.js]: Addressed... 26545edf

**Comment by jonambas** [comment]
`verifyEmailToken` in `currentUser` can be silent - although it is a wrapper around another action that should not be silent

**Comment by bkemper** [comment]
Rebased.

File: src/actions/currentUser.js [hunk 14]
```diff
@@ -46,7 +46,8 @@ export function verifyEmail(data = {}, type = 'VERIFY_EMAIL') {
       meta: {
         method: 'POST',
         url: `/users/${username}/verify`,
-        data
+        data,
+        showErrorAlert: false
```
Comment (jonambas) [inline src/actions/currentUser.js]: this should be in the error below `verifyEmailToken`

Comment (bkemper) [reply src/actions/currentUser.js]: Addressed... e835de7

**Pull request merged**